### PR TITLE
Dev 2.0

### PR DESCRIPTION
**Pull request by simeongr**
Changed algorithm. Using reversed polish notation.

### CODE REVIEW

File: Calculator/src/main/java/com/sap/calculator/Expression.java [hunk 1]
```diff
@@ -17,26 +21,62 @@ protected static String removeSpaces(String expression) throws InvalidExpression
 		return expression.replaceAll(" ", "");
 	}
 	
-	public static String[] expressionToArray(String expression) throws InvalidExpressionException {
-		String[] result = getNumbersAndOperators(removeSpaces(expression));
-		checkValidity(result);
+	public static List<String> expressionToArray(String expression) throws InvalidExpressionException {
+		List<String> result = handleNegativeNumbers(/*Arrays.asList(*/getNumbersAndOperators(removeSpaces(expression)));
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/Expression.java]: Remove comments

File: Calculator/src/main/java/com/sap/calculator/RpnConverter.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.Queue;
+import java.util.Stack;
+
+import com.sap.calculator.exceptions.InvalidExpressionException;
+
+public class RpnConverter {
+	
+	private Stack<String> stack = new Stack<>();
+	private Queue<String> queue = new LinkedList<>();
+	private Map<String, Integer> operatorsValue = new HashMap<>();
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/RpnConverter.java]: Add generic types in the initializaiton of the HashMap - the IDE should suggest that already

File: Calculator/src/main/java/com/sap/calculator/RpnConverter.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+				while (!stack.isEmpty() && operatorsValue.get(symbol) <= operatorsValue.get(stack.peek())) {
+					queue.add(stack.pop());
+				}
+				stack.add(symbol);
+			}
+			else {
+				if (Expression.isOperand(symbol))
+					queue.add(symbol);
+				else
+					throw new InvalidExpressionException("Invalid expression.");
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/RpnConverter.java]: you already do extensive validation. can this happen?

File: Calculator/src/main/java/com/sap/calculator/RpnConverter.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+			}
+		}
+		while(!stack.isEmpty()) {
+			queue.add(stack.pop());
+		}
+		
+		return new ArrayList<>(queue);
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/RpnConverter.java]: Again - specify the type of the list

File: Calculator/src/main/java/com/sap/calculator/Calculator.java [hunk 5]
```diff
@@ -1,45 +1,46 @@
...
-				result *= Double.parseDouble(nextOperand);
-				break;
-
-			case "/":
-				result /= Double.parseDouble(nextOperand);
-				break;
-			default:
-				throw new InvalidExpressionException("The expression is not valid.");
+		RpnConverter rpnConverter = new RpnConverter();
+		List<String> test = Expression.expressionToArray(expression);
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/Calculator.java]: Do you need that? Seems like dead code

File: Calculator/src/main/java/com/sap/calculator/Calculator.java [hunk 6]
```diff
@@ -1,45 +1,46 @@
...
+						
+						if (symbol.equals("+"))
+							stack.add(first + second);
+						if (symbol.equals("-"))
+							stack.add(first - second);
+						if (symbol.equals("*"))
+							stack.add(first * second);
+						if (symbol.equals("/"))
+							stack.add(first / second);
+					} catch (Exception ex) {
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/Calculator.java]: When can this happen? I don't see any code throwing exception here.

File: Calculator/src/main/java/com/sap/calculator/Calculator.java [hunk 7]
```diff
@@ -1,45 +1,46 @@
...
+						throw new InvalidExpressionException("Invalid expression.");
+					}
+				}
+				else
```
Comment (HristoDobchev) [inline Calculator/src/main/java/com/sap/calculator/Calculator.java]: Again - validation was done so this does not seem to be correct.

**Comment by HristoDobchev** [comment]
The code is readable but you make too many iterations and check the same things too many times (e.g. isOperator isOperand etc.) Try to preserve the information that you already determined. You can do that by introducing another structure(class) after parsing that can save the contextual information.

File: Calculator/src/main/java/com/sap/calculator/Calculator.java (hunk 6 above)
Comment (simeongr) [reply Calculator/src/main/java/com/sap/calculator/Calculator.java]: When you have incorrect expression like this one "1+3++4//", the rpn looks like this [1, 3, +, +, 4, +]. At the second "+" you have only one operand in the stack, so it throws empty stack exception. This is validation process, i think its needed.

**Pull request merged**